src/concrete/ml/onnx/ops_impl.py

```diff
@@ -1495,7 +1495,12 @@ def numpy_cast(data: numpy.ndarray, *, to: int) -> Tuple[numpy.ndarray]:
     if isinstance(data, RawOpOutput):
         return (data.astype(onnx.helper.tensor_dtype_to_np_dtype(to)).view(RawOpOutput),)
 
-    assert_true(to == onnx.TensorProto.BOOL)
+    allowed_types = (onnx.TensorProto.BOOL, onnx.TensorProto.FLOAT, onnx.TensorProto.DOUBLE)
+    assert to in allowed_types, (
+        f"Invalid 'to' data type: {onnx.TensorProto.DataType.Name(to)}. "
+        f"Only {', '.join(onnx.TensorProto.DataType.Name(t) for t in allowed_types)}"
+        "are allowed for casting."
+    )
 
     # Will be used for traced values
     return (data.astype(numpy.float64),)
```

Comment:
not sure what is happening here then : we check that `to` is a bool, float or double, but we only cast to double without talking `to` into account ?

Reply:
Yes it's pretty weird indeed. And before it was a cast to bool and then we casted to float64 so even weirder.

I think the idea is that we want to be able to support cast (i.e. not break because of a cast in the torch forward) but we don't want to change the FHE semantic. Casting to anything else than float64 is dangerous as it could be interpreted by concrete-python improperly and close a LUT.

Actually as @andrei-stoian-zama stated casting to float64 could be potentially dangerous as well if inputs where integers as it would mean opening a LUT (while the cast actually as no real effect in FHE).

Reply:
ah yes actually now that I think of it, I remember already having some issues with that `numpy_cast` in the past

it feels like there's something we should do about your last point though !

Reply:
But would this kind of check integer be for the numpy cast only? I think this is a more general concern. Also we have function that do work on integers (tree specific numpy method) and some function that work on both (numpy_gemm).

Reply:
yes yes for sure this is more of a general concern, maybe we should summarize this in a dedicated issue ?

Reply:
Will do

Reply:
We could maybe have something that does:
```
if not int: return float
if int: return int
```
or smth like that